core/src/main/java/com/linecorp/armeria/internal/DefaultTimeoutController.java

```diff
@@ -76,100 +84,166 @@ public void setTimeoutTask(TimeoutTask timeoutTask) {
-            final long passedTimeMillis = TimeUnit.NANOSECONDS.toMillis(currentNanoTime - lastStartTimeNanos);
-            final long remainingTimeoutMillis = LongMath.saturatedSubtract(timeoutMillis, passedTimeMillis);
-            lastStartTimeNanos = currentNanoTime;
-            if (remainingTimeoutMillis == newTimeoutMillis) {
-                return;
-            }
+        // Cancel the previously scheduled timeout, if exists.
+        cancelTimeout();
+        timeoutMillis = newTimeoutMillis;
+        timeoutFuture = eventLoop.schedule(this::invokeTimeoutTask, newTimeoutMillis, TimeUnit.MILLISECONDS);
+        state = State.SCHEDULED;
+        return true;
+    }
 
-            // Cancel the previously scheduled timeout, if exists.
-            cancelTimeout();
-            timeoutMillis = newTimeoutMillis;
-            timeoutFuture = eventLoop.schedule(timeoutTask, newTimeoutMillis,
-                                               TimeUnit.MILLISECONDS);
+    /**
+     * Trigger the current timeout immediately.
+     *
+     * @return {@code true} if the current timeout is triggered successfully.
+     *         {@code false} if no timeout was scheduled previously, the timeout has been triggered already
+     *         or the {@link TimeoutTask} could not be scheduled now.
+     */
+    @Override
+    public boolean timeoutNow() {
+        checkState(timeoutTask != null,
+                   "setTimeoutTask(timeoutTask) is not called yet.");
+
+        if (State.SCHEDULED != state || !timeoutTask.canSchedule()) {
+            return false;
         }
+
+        if (cancelTimeout()) {
+            invokeTimeoutTask();
+            return true;
+        }
+
+        return false;
     }
 
     @Override
     public boolean cancelTimeout() {
-        final ScheduledFuture<?> timeoutFuture = this.timeoutFuture;
-        if (timeoutFuture == null) {
+        if (State.TIMED_OUT == state) {
+            return false;
+        }
+        if (State.DISABLED == state) {
             return true;
         }
 
+        final ScheduledFuture<?> timeoutFuture = this.timeoutFuture;
+        assert timeoutFuture != null;
+
         this.timeoutFuture = null;
-        return timeoutFuture.cancel(false);
+        final boolean canceled = timeoutFuture.cancel(false);
+        state = State.DISABLED;
+        return canceled;
     }
 
     private void ensureInitialized() {
         checkState(timeoutTask != null,
                    "setTimeoutTask(timeoutTask) is not called yet.");
-        checkState(firstStartTimeNanos > 0,
-                   "initTimeout(timeoutMillis) is not called yet.");
+        if (firstExecutionTimeNanos == 0) {
+            firstExecutionTimeNanos = lastExecutionTimeNanos = System.nanoTime();
+        }
+    }
+
+    @Nullable
+    private void invokeTimeoutTask() {
+        if (timeoutTask != null) {
+            timeoutTask.run();
+            state = State.TIMED_OUT;
```

Comment:
Could set in `finally` block, just in case `run()` throws an exception

Reply:
Nice tip, thanks!